src/components/NcHeaderMenu/NcHeaderMenu.vue

```diff
@@ -0,0 +1,315 @@
+  - along with this program.  If not, see <http://www.gnu.org/licenses/>.
+  -
+  -->
+
+<docs>
+This component is made to be used in the Nextcloud top header.
+
+```
+<template>
+	<div id="nextcloud-header">
+		<NcHeaderMenu id="search"
+			aria-label="Search">
+			<template #trigger>
+				<Magnify />
+			</template>
+			<div>
+				<input placeholder="Search for files, comments, contacts..." type="search" style="width: 99%;" />
+				<NcEmptyContent
+					title="Search"
+					description="Start typing to search">
+					<template #icon>
+						<Magnify />
+					</template>
+				</NcEmptyContent>
+			</div>
+		</NcHeaderMenu>
+	</div>
+</template>
+<script>
+import Magnify from 'vue-material-design-icons/Magnify'
+
+export default {
+	components: {
+		Magnify,
+	},
+}
+</script>
+<style>
+#nextcloud-header {
+	display: flex;
+	justify-content: right;
+	background-color: var(--color-primary);
+}
+</style>
+```
+</docs>
+
+<template>
+	<div :id="id"
+		v-click-outside="clickOutsideConfig"
+		:class="{ 'header-menu--opened': opened }"
+		class="header-menu">
+		<!-- Open trigger icon -->
+		<a class="header-menu__trigger"
+			href="#"
+			:aria-label="ariaLabel"
+			:aria-controls="`header-menu-${id}`"
+			:aria-expanded="opened.toString()"
+			@click.prevent="toggleMenu">
+			<!-- @slot Icon trigger slot. Make sure the svg path
+				is at least 16px. Usually mdi icon works at 20px -->
+			<slot name="trigger" />
+		</a>
+
+		<!-- Visual triangle -->
+		<div v-show="opened" class="header-menu__carret" />
+
+		<!-- Menu opened content -->
+		<div v-show="opened"
+			:id="`header-menu-${id}`"
+			class="header-menu__wrapper"
+			role="menu">
+			<div class="header-menu__content">
+				<!-- @slot Main content -->
+				<slot />
+			</div>
+		</div>
+	</div>
+</template>
+
+<script>
+import { directive as ClickOutside } from 'v-click-outside'
+import excludeClickOutsideClasses from '../../mixins/excludeClickOutsideClasses/index.js'
+
+export default {
+	name: 'NcHeaderMenu',
+
+	directives: {
+		ClickOutside,
+	},
+
+	mixins: [
+		excludeClickOutsideClasses,
+	],
+
+	props: {
+		/**
+		 * Unique id for this menu
+		 */
+		id: {
+			type: String,
+			required: true,
+		},
+
+		/**
+		 * aria-label attribute of the menu open button
+		 */
+		ariaLabel: {
+			type: String,
+			default: '',
+		},
+
+		/**
+		 * Current menu open state
+		 */
+		open: {
+			type: Boolean,
+			default: false,
+		},
+	},
+
+	emits: [
+		'close',
+		'open',
+		'update:open',
+		'cancel',
+	],
+
+	data() {
+		return {
+			opened: this.open,
+			clickOutsideConfig: {
+				handler: this.closeMenu,
+				middleware: this.clickOutsideMiddleware,
+			},
+			shortcutsDisabled: window.OCP?.Accessibility?.disableKeyboardShortcuts?.(),
+		}
+	},
+
+	watch: {
+		open(newVal) {
+			this.opened = newVal
+			this.$nextTick(() => {
+				if (this.opened) {
+					this.openMenu()
+				} else {
+					this.closeMenu()
+				}
+			})
```

Comment:
Why do we need to set `this.opened` when `openMenu` or `closeMenu` will do it again later ?
Would this work ?
```js
if (newVal) {
	this.openMenu()
} else {
	this.closeMenu()
}
```

Reply:
Because we keep a local state `opened`.
The prop is `open`. If we don't do that, it means the dev _have_ to provide a writeable prop (with `.sync` for example)

Like this it means the dev can manually trigger open/close from the outside or not provide it and let the component handle itself.

If the external `open` prop change, we trigger the same way as if the user would have opened or close the menu.
Changing the internal data variable `opened`

Reply:
I added comments, it's to properly fire the events too. But we make sure the component is updated before doing so

Reply:
https://github.com/nextcloud/server/blob/abaa3ef61b9a6177ca3fe965e837e6e9bca16a4b/core/src/views/UnifiedSearch.vue#L371

Here is an example. If we trigger too soon, we cannot focus the input.

Reply:
Then `nextTick` needs to go in `openMenu`/`closeMenu` if we want to be consistent, no ?

Because here, I understand that the parent will receive the event before the next tick if it comes from here: https://github.com/nextcloud/nextcloud-vue/pull/3489/files/9c21762167bba03dd398cd3d5a0cc7a58b9e4f5d#diff-7af842de9d50674691572f5571b0849046baa1f66655a53bc2d1d6f51fd14ddcR77

Reply:
Might be yeah :shrug:

Reply:
Do we want to properly do the `open` and `opened` differences in emitted events ?

Reply:
Done